src/test/util/setup_common.h

```diff
@@ -13,6 +13,7 @@
 #include <pubkey.h>
 #include <random.h>
 #include <stdexcept>
+#include <test/util/net.h>
```

Comment:
I tried extracting the mocked sockets out from `test/util/net.h` into a separate `test/util/mocked_sockets.h` and then compared the re-compilation time if that is modified (e.g. something in `DynSock` declaration is changed). In both cases recompilation takes ~28 seconds, so it makes no difference (in practice).

Here is the change (on top of this PR):

<details>
<summary>mocked_sockets.h</summary>

```diff
commit 398a2d5717b6ef6760e48eac92aa92127157a67a (HEAD -> e2e_tests)
Parent: 4557cc336fd8eb321b0db024b70213f46017071c
Author:     Vasil Dimov <vd@FreeBSD.org>
AuthorDate: Wed Jun 28 09:10:41 2023 +0200
Commit:     Vasil Dimov <vd@FreeBSD.org>
CommitDate: Wed Jun 28 09:10:41 2023 +0200
gpg: Signature made Wed Jun 28 09:11:03 2023 CEST
gpg:                using RSA key E64D8D45614DB07545D9CCC154DF06F64B55CBBF
gpg: Good signature from "Vasil Dimov <vd@myforest.net>" [ultimate]
gpg:                 aka "Vasil Dimov <vd@FreeBSD.org>" [ultimate]
gpg:                 aka "Vasil Dimov <vasild@gmail.com>" [ultimate]


    wip: mocked_sockets.h

diff --git a/src/test/util/net.h b/src/test/util/mocked_sockets.h
similarity index 67%
copy from src/test/util/net.h
copy to src/test/util/mocked_sockets.h
index b5ab7699f0..cdf6cb5d96 100644
--- a/src/test/util/net.h
+++ b/src/test/util/mocked_sockets.h
@@ -1,111 +1,27 @@
-// Copyright (c) 2020-2022 The Bitcoin Core developers
+// Copyright (c) 2023-2023 The Bitcoin Core developers
 // Distributed under the MIT software license, see the accompanying
 // file COPYING or http://www.opensource.org/licenses/mit-license.php.
 
-#ifndef BITCOIN_TEST_UTIL_NET_H
-#define BITCOIN_TEST_UTIL_NET_H
+#ifndef BITCOIN_TEST_UTIL_MOCKED_SOCKETS_H
+#define BITCOIN_TEST_UTIL_MOCKED_SOCKETS_H
 
 #include <compat/compat.h>
 #include <net.h>
-#include <netaddress.h>
 #include <netmessagemaker.h>
-#include <node/eviction.h>
-#include <util/sock.h>
 #include <sync.h>
+#include <util/sock.h>
+#include <version.h>
 
-#include <array>
-#include <cassert>
 #include <chrono>
-#include <condition_variable>
-#include <cstring>
+#include <cstdint>
 #include <memory>
 #include <optional>
 #include <string>
 #include <vector>
 
-struct ConnmanTestMsg : public CConnman {
-    using CConnman::CConnman;
-
-    void SetPeerConnectTimeout(std::chrono::seconds timeout)
-    {
-        m_peer_connect_timeout = timeout;
-    }
-
-    void AddTestNode(CNode& node)
-    {
-        LOCK(m_nodes_mutex);
-        m_nodes.push_back(&node);
-    }
-    void ClearTestNodes()
-    {
-        LOCK(m_nodes_mutex);
-        for (CNode* node : m_nodes) {
-            delete node;
-        }
-        m_nodes.clear();
-    }
-
-    void Handshake(CNode& node,
-                   bool successfully_connected,
-                   ServiceFlags remote_services,
-                   ServiceFlags local_services,
-                   int32_t version,
-                   bool relay_txs)
-        EXCLUSIVE_LOCKS_REQUIRED(NetEventsInterface::g_msgproc_mutex);
-
-    void ProcessMessagesOnce(CNode& node) EXCLUSIVE_LOCKS_REQUIRED(NetEventsInterface::g_msgproc_mutex) { m_msgproc->ProcessMessages(&node, flagInterruptMsgProc); }
-
-    void NodeReceiveMsgBytes(CNode& node, Span<const uint8_t> msg_bytes, bool& complete) const;
-
-    bool ReceiveMsgFrom(CNode& node, CSerializedNetMsg& ser_msg) const;
-};
-
-constexpr ServiceFlags ALL_SERVICE_FLAGS[]{
-    NODE_NONE,
-    NODE_NETWORK,
-    NODE_BLOOM,
-    NODE_WITNESS,
-    NODE_COMPACT_FILTERS,
-    NODE_NETWORK_LIMITED,
-};
-
-constexpr NetPermissionFlags ALL_NET_PERMISSION_FLAGS[]{
-    NetPermissionFlags::None,
-    NetPermissionFlags::BloomFilter,
-    NetPermissionFlags::Relay,
-    NetPermissionFlags::ForceRelay,
-    NetPermissionFlags::NoBan,
-    NetPermissionFlags::Mempool,
-    NetPermissionFlags::Addr,
-    NetPermissionFlags::Download,
-    NetPermissionFlags::Implicit,
-    NetPermissionFlags::All,
-};
-
-constexpr ConnectionType ALL_CONNECTION_TYPES[]{
-    ConnectionType::INBOUND,
-    ConnectionType::OUTBOUND_FULL_RELAY,
-    ConnectionType::MANUAL,
-    ConnectionType::FEELER,
-    ConnectionType::BLOCK_RELAY,
-    ConnectionType::ADDR_FETCH,
-};
-
-constexpr auto ALL_NETWORKS = std::array{
-    Network::NET_UNROUTABLE,
-    Network::NET_IPV4,
-    Network::NET_IPV6,
-    Network::NET_ONION,
-    Network::NET_I2P,
-    Network::NET_CJDNS,
-    Network::NET_INTERNAL,
-};
-
-std::vector<NodeEvictionCandidate> GetRandomNodeEvictionCandidates(int n_candidates, FastRandomContext& random_context);
-
 /**
  * A mocked Sock alternative that succeeds on all operations.
  * Returns infinite amount of 0x0 bytes on reads.
  */
 class ZeroSock : public Sock
 {
@@ -257,7 +173,8 @@ void DynSock::Pipe::PushNetMsg(const std::string& type, Args&&... payload)
     LOCK(m_mutex);
     m_data.insert(m_data.end(), header.begin(), header.end());
     m_data.insert(m_data.end(), msg.data.begin(), msg.data.end());
     m_cond.notify_all();
 }
 
-#endif // BITCOIN_TEST_UTIL_NET_H
+#endif // BITCOIN_TEST_UTIL_MOCKED_SOCKETS_H
+
diff --git a/src/test/util/net.cpp b/src/test/util/net.cpp
index ac866d7c89..1acdbf7061 100644
--- a/src/test/util/net.cpp
+++ b/src/test/util/net.cpp
@@ -8,12 +8,13 @@
 #include <node/eviction.h>
 #include <net.h>
 #include <net_processing.h>
 #include <netmessagemaker.h>
 #include <span.h>
 #include <sync.h>
+#include <test/util/mocked_sockets.h>
 
 #include <chrono>
 #include <optional>
 #include <vector>
 
 void ConnmanTestMsg::Handshake(CNode& node,
diff --git a/src/test/util/net.h b/src/test/util/net.h
index b5ab7699f0..7bad1ee4af 100644
--- a/src/test/util/net.h
+++ b/src/test/util/net.h
@@ -100,164 +100,7 @@ constexpr auto ALL_NETWORKS = std::array{
     Network::NET_CJDNS,
     Network::NET_INTERNAL,
 };
 
 std::vector<NodeEvictionCandidate> GetRandomNodeEvictionCandidates(int n_candidates, FastRandomContext& random_context);
 
-/**
- * A mocked Sock alternative that succeeds on all operations.
- * Returns infinite amount of 0x0 bytes on reads.
- */
-class ZeroSock : public Sock
-{
-public:
-    ZeroSock();
-
-    ~ZeroSock() override;
-
-    ssize_t Send(const void*, size_t len, int) const override;
-
-    ssize_t Recv(void* buf, size_t len, int flags) const override;
-
-    int Connect(const sockaddr*, socklen_t) const override;
-
-    int Bind(const sockaddr*, socklen_t) const override;
-
-    int Listen(int) const override;
-
-    std::unique_ptr<Sock> Accept(sockaddr* addr, socklen_t* addr_len) const override;
-
-    int GetSockOpt(int level, int opt_name, void* opt_val, socklen_t* opt_len) const override;
-
-    int SetSockOpt(int, int, const void*, socklen_t) const override;
-
-    int GetSockName(sockaddr* name, socklen_t* name_len) const override;
-
-    bool SetNonBlocking() const override;
-
-    bool IsSelectable() const override;
-
-    bool Wait(std::chrono::milliseconds timeout,
-              Event requested,
-              Event* occurred = nullptr) const override;
-
-    bool WaitMany(std::chrono::milliseconds timeout, EventsPerSock& events_per_sock) const override;
-
-private:
-    ZeroSock& operator=(Sock&& other) override;
-};
-
-/**
- * A mocked Sock alternative that returns a statically contained data upon read and succeeds
- * and ignores all writes. The data to be returned is given to the constructor and when it is
- * exhausted an EOF is returned by further reads.
- */
-class StaticContentsSock : public ZeroSock
-{
-public:
-    explicit StaticContentsSock(const std::string& contents);
-
-    /**
-     * Return parts of the contents that was provided at construction until it is exhausted
-     * and then return 0 (EOF).
-     */
-    ssize_t Recv(void* buf, size_t len, int flags) const override;
-
-private:
-    StaticContentsSock& operator=(Sock&& other) override;
-
-    const std::string m_contents;
-    mutable size_t m_consumed{0};
-};
-
-/**
- * A mocked Sock alternative that allows providing the data to be returned by Recv()
- * and inspecting the data that has been supplied to Send().
- */
-class DynSock : public ZeroSock
-{
-public:
-    /**
-     * Unidirectional bytes or CNetMessage queue (FIFO).
-     */
-    class Pipe
-    {
-    public:
-        /**
-         * Get bytes and remove them from the pipe.
-         * @param[in] buf Destination to write bytes to.
-         * @param[in] len Write up to this number of bytes.
-         * @param[in] flags Same as the flags of `recv(2)`. Just `MSG_PEEK` is honored.
-         * @return The number of bytes written to `buf`. `0` if `Eof()` has been called.
-         * If no bytes are available then `-1` is returned and `errno` is set to `EAGAIN`.
-         */
-        ssize_t GetBytes(void* buf, size_t len, int flags = 0) EXCLUSIVE_LOCKS_REQUIRED(!m_mutex);
-
-        /**
-         * Deserialize a `CNetMessage` and remove it from the pipe.
-         * If not enough bytes are available then the function will wait. If parsing fails
-         * or EOF is signaled to the pipe, then `std::nullopt` is returned.
-         */
-        std::optional<CNetMessage> GetNetMsg() EXCLUSIVE_LOCKS_REQUIRED(!m_mutex);
-
-        /**
-         * Push bytes to the pipe.
-         */
-        void PushBytes(const void* buf, size_t len) EXCLUSIVE_LOCKS_REQUIRED(!m_mutex);
-
-        /**
-         * Construct and push CNetMessage to the pipe.
-         */
-        template <typename... Args>
-        void PushNetMsg(const std::string& type, Args&&... payload) EXCLUSIVE_LOCKS_REQUIRED(!m_mutex);
-
-        /**
-         * Signal end-of-file on the receiving end (`GetBytes()` or `GetNetMsg()`).
-         */
-        void Eof() EXCLUSIVE_LOCKS_REQUIRED(!m_mutex);
-
-    private:
-        Mutex m_mutex;
-        std::condition_variable m_cond;
-        std::vector<uint8_t> m_data GUARDED_BY(m_mutex);
-        bool m_eof GUARDED_BY(m_mutex){false};
-    };
-
-    struct Pipes {
-        Pipe recv;
-        Pipe send;
-    };
-
-    explicit DynSock(std::shared_ptr<Pipes> pipes);
-
-    ~DynSock();
-
-    ssize_t Recv(void* buf, size_t len, int flags) const override;
-
-    ssize_t Send(const void* buf, size_t len, int) const override;
-
-    bool Wait(std::chrono::milliseconds timeout,
-              Event requested,
-              Event* occurred = nullptr) const override;
-
-    bool WaitMany(std::chrono::milliseconds timeout, EventsPerSock& events_per_sock) const override;
-
-private:
-    DynSock& operator=(Sock&&) override;
-
-    std::shared_ptr<Pipes> m_pipes;
-};
-
-template <typename... Args>
-void DynSock::Pipe::PushNetMsg(const std::string& type, Args&&... payload)
-{
-    auto msg = CNetMsgMaker(INIT_PROTO_VERSION).Make(type, std::forward<Args>(payload)...);
-    std::vector<unsigned char> header;
-    V1TransportSerializer().prepareForTransport(msg, header);
-
-    LOCK(m_mutex);
-    m_data.insert(m_data.end(), header.begin(), header.end());
-    m_data.insert(m_data.end(), msg.data.begin(), msg.data.end());
-    m_cond.notify_all();
-}
-
 #endif // BITCOIN_TEST_UTIL_NET_H
diff --git a/src/test/util/setup_common.h b/src/test/util/setup_common.h
index d3137fa69e..1daac4a57c 100644
--- a/src/test/util/setup_common.h
+++ b/src/test/util/setup_common.h
@@ -10,13 +10,13 @@
 #include <node/caches.h>
 #include <node/context.h> // IWYU pragma: export
 #include <primitives/transaction.h>
 #include <pubkey.h>
 #include <random.h>
 #include <stdexcept>
-#include <test/util/net.h>
+#include <test/util/mocked_sockets.h>
 #include <util/chaintype.h>
 #include <util/check.h>
 #include <util/fs.h>
 #include <util/string.h>
 #include <util/vector.h>
 
```
</details>

Or did you mean something else?

Reply:
```
PR:
real    0m29.623s
user    10m58.337s
sys     0m51.187s

real    0m29.829s
user    10m58.359s
sys     0m50.912s

mocked_sockets.h:
real    0m29.528s
user    11m12.208s
sys     1m0.406s

real    0m28.831s
user    10m45.615s
sys     0m49.680s
```

I guess the "total CPU time" is real+sys from above. It all seems within noise. In both cases all tests would be recompiled, maybe I got @jonatack's comment wrong?

Reply:
Alright, so I got it the other way around. So the point is to avoid recompiling all tests if e.g. `ALL_NETWORKS` is changed (defined in `test/util/net.h`). That lowers the re-compilation from 28-29 sec to about 10 sec (if `src/test/util/net.h` is modified).

I will leave it as it is because moving the code around would further increase the size of this PR which suffers from lack of interest from reviewers and I think expanding it may further turn people away.